source/common/sequencer_impl.cc

```diff
@@ -82,49 +81,50 @@ void SequencerImpl::updateStartBlockingTimeIfNeeded() {
 
 void SequencerImpl::run(bool from_periodic_timer) {
   ASSERT(running_);
+  // Update time so that our time source will yield an up-to-date monotonic time sample to work with
+  // during this function call.
+  dispatcher_.updateApproximateMonotonicTime();
```

Comment:
This feels a little bit counter intuitive. We now have a clock (time) abstraction that moves forward only when we tell it to. This might lead to programming errors and the reasoning isn't obvious?

If I understand fixes in this PR correctly, when compared to the first attempt on phases we are now passing the expected start time to all components. Why are we also caching time? Is it an unrelated performance improvement or required for phases? If it isn't required for phases, can we handle it in a separate PR and start by discussing the numbers that lead us to believe that caching is necessary?

Reply:
I will back this out and create a separate PR. It’s not needed to address the problem I noticed, but rather a separate potential enhancement. For context, performance is not the goal of the caching here, but a side effect:, it’s consistency of “now” across the sequencer, rate limiters and termination predicates that I liked.  I thought this might be nice as it avoids subtle edge cases caused by small slight differences caused by querying the clock independently.

Reply:
Back out in 048dd15daeed78a068d37e94b8d00c96948eab5d